-1

**Failed Tests**:   Build  ClangBuild 
**Size**: This PR adds an extra 32KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-99736e/45679/summary.html
**COMMIT**: 707ec7f4397178d662ee4073e906f9d29ee25db4
**CMSSW**: CMSSW_15_1_X_2025-04-22-2300/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47931/45679/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>Copying tmp/el8_amd64_gcc12/src/HeterogeneousCore/AlpakaInterface/test/alpakaWorkDivisionROCmAsync/libalpakaWorkDivisionROCmAsync_rocm.a to productstore area:
Copying tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/plugins/RecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync/libRecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync_rocm.a to productstore area:
cp: cannot stat 'tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/plugins/RecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync/libRecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync_rocm.a': No such file or directory
Copying tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/test/deviceVertexFinderByDensity_tROCmAsync/libdeviceVertexFinderByDensity_tROCmAsync_rocm.a to productstore area:
>> Deleted: tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/plugins/RecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync/libRecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync_rocm.a
gmake: *** [config/SCRAM/GMake/Makefile.rules:1886: tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/plugins/RecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync/libRecoPixelVertexingPixelVertexFindingPluginsPortableROCmAsync_rocm.a] Error 1
cp: cannot stat 'tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/test/deviceVertexFinderByDensity_tROCmAsync/libdeviceVertexFinderByDensity_tROCmAsync_rocm.a': No such file or directory
>> Deleted: tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/test/deviceVertexFinderByDensity_tROCmAsync/libdeviceVertexFinderByDensity_tROCmAsync_rocm.a
gmake: *** [config/SCRAM/GMake/Makefile.rules:1886: tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/test/deviceVertexFinderByDensity_tROCmAsync/libdeviceVertexFinderByDensity_tROCmAsync_rocm.a] Error 1
Copying tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/test/deviceVertexFinderDBSCAN_tROCmAsync/libdeviceVertexFinderDBSCAN_tROCmAsync_rocm.a to productstore area:
cp: cannot stat 'tmp/el8_amd64_gcc12/src/RecoVertex/PixelVertexFinding/test/deviceVertexFinderDBSCAN_tROCmAsync/libdeviceVertexFinderDBSCAN_tROCmAsync_rocm.a': No such file or directory
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Creating project symlinks
>> Entering Package HeterogeneousCore/AlpakaInterface
>> Entering Package RecoTracker/PixelSeeding
>> Entering Package RecoVertex/PixelVertexFinding
>> Compile sequence completed for CMSSW CMSSW_15_1_X_2025-04-22-2300
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
Command exited with non-zero status 1
	Command being timed: "scram build -k -j 32 COMPILER=llvm compile BUILD_LOG=yes"
	User time (seconds): 2819.45
	System time (seconds): 257.49
	Percent of CPU this job got: 649%
</pre>

--- a/HeterogeneousCore/AlpakaInterface/interface/debug.h
+++ b/HeterogeneousCore/AlpakaInterface/interface/debug.h
@@ -1,0 +1,25 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_debug_h
+#define HeterogeneousCore_AlpakaInterface_interface_debug_h
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE::debug {
+
+  [[nodiscard]] ALPAKA_FN_ACC inline bool always_true() {
+    volatile const bool flag = true;
+    return flag;
+  }
+
+  [[nodiscard]] ALPAKA_FN_ACC inline bool always_false() {
+    volatile const bool flag = false;
+    return flag;
+  }
+
+  ALPAKA_FN_ACC inline void do_not_optimise(const auto& arg) {
+    volatile const auto* ptr = &arg;
+    (void)*ptr;
+  }
+
+}  // namespace ALPAKA_ACCELERATOR_NAMESPACE::debug
+
+#endif  // HeterogeneousCore_AlpakaInterface_interface_debug_h

--- a/RecoTracker/PixelSeeding/plugins/alpaka/BrokenLineFit.dev.cc
+++ b/RecoTracker/PixelSeeding/plugins/alpaka/BrokenLineFit.dev.cc
@@ -7,6 +7,7 @@
 
 #include "DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsSoA.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/debug.h"
 #include "RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h"
 #include "RecoTracker/PixelTrackFitting/interface/alpaka/BrokenLine.h"
 
@@ -182,6 +183,9 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
                                   double *__restrict__ phits,
                                   float *__restrict__ phits_ge,
                                   double *__restrict__ pfast_fit) const {
+      // workaround for #47808
+      debug::do_not_optimise(results_view);
+
       ALPAKA_ASSERT_ACC(results_view.pt());
       ALPAKA_ASSERT_ACC(results_view.eta());
       ALPAKA_ASSERT_ACC(results_view.chi2());

--- a/RecoVertex/PixelVertexFinding/plugins/alpaka/clusterTracksByDensity.h
+++ b/RecoVertex/PixelVertexFinding/plugins/alpaka/clusterTracksByDensity.h
@@ -10,6 +10,7 @@
 #include "DataFormats/VertexSoA/interface/ZVertexSoA.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/debug.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/warpsize.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
 #include "RecoVertex/PixelVertexFinding/interface/PixelVertexWorkSpaceLayout.h"
@@ -31,6 +32,9 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::vertexFinder {
                                                              float errmax,  // max error to be "seed"
                                                              float chi2max  // max normalized distance to cluster
   ) {
+    // workaround for #47808
+    debug::do_not_optimise(ws);
+
     constexpr bool verbose = false;
 
     if constexpr (verbose) {

--- a/RecoVertex/PixelVertexFinding/plugins/alpaka/clusterTracksDBSCAN.h
+++ b/RecoVertex/PixelVertexFinding/plugins/alpaka/clusterTracksDBSCAN.h
@@ -10,6 +10,7 @@
 #include "DataFormats/VertexSoA/interface/ZVertexSoA.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/HistoContainer.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/debug.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
 #include "RecoVertex/PixelVertexFinding/interface/PixelVertexWorkSpaceLayout.h"
 
@@ -30,6 +31,9 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::vertexFinder {
                                   float errmax,  // max error to be "seed"
                                   float chi2max  // max normalized distance to cluster
     ) const {
+      // workaround for #47808
+      debug::do_not_optimise(ws);
+
       constexpr bool verbose = false;
 
       if constexpr (verbose) {

